File: go/problems/94_binary_tree_inorder_traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	val := []int{}
	left := inorderTraversal(root.Left)
	val = append(val, left...)

	val = append(val, root.Val)

	right := inorderTraversal(root.Right)
	val = append(val, right...)
	return val
}

File: ts/problems/94-binary-tree-inorder-traversal.ts
export type TreeNode = {
  val: number;
  left: TreeNode;
  right: TreeNode;
} | null;

export function inorderTraversal(root: TreeNode | null): number[] {
  if (!root) return [];

  const val: number[] = [];
  val.push(...inorderTraversal(root.left));
  val.push(root.val);
  val.push(...inorderTraversal(root.right));

  return val;
}
